### PR TITLE
Row class

### PR DESCRIPTION
**Pull request by josephShokry**

### CODE REVIEW

File: source/models/row.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        if value is not None:
+            self.__init_by_value(value)
+        elif primary_key is not None:
+            self.__init_by_primary_key()
+        else:
+            raise WrongInput(message = "the value and PK of the row is null")
+        
+    def __init_by_value(self, value):
+        self.value = value
+        self.primary_key = value[self.table.table_metadata.primary_key]
```
Comment (mohamedabdelsalam) [inline source/models/row.py:12]: In this case, primary key can be null. I would suggest set primary_key with random variable if it is null and update data[primary_key]

File: source/models/row.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    def __init_by_primary_key(self,primary_key):
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: Why don’t you have a static method 
Row.load(primaru_key)
- read from file 
- New Row(data)

File: source/models/row.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            self.value[self.table.table_metadata.primary_key] = str(uuid.uuid4().node)
+        row_json_data = json.dumps(self.value, indent = 2)
+        with open(self.get_path(), 'x') as row_file: 
+            row_file.write(row_json_data)
+        self.update_index()
+
+    def row_exists(self):
+        return os.path.isfile(self.get_path())
+
+    def get_path(self):
```
Comment (mohamedabdelsalam) [inline source/models/row.py:28]: All rows should be in folder data

in table folder there is two folders : -
Lock
Data

File: source/models/row.py (hunk 1 above)
Comment (josephShokry) [reply source/models/row.py:12]: I did so in the set command. so I suppose that the entered value is containing the PK if it is a given one or a random one,
so why it is preferred to be here?

File: source/models/row.py (hunk 2 above)
Comment (josephShokry) [reply source/models/row.py]: if I understood ```__get_by_primary_key``` do so.

File: source/models/row.py (hunk 1 above)
Comment (josephShokry) [reply source/models/row.py:12]: ok done

File: source/models/row.py (hunk 2 above)
Comment (josephShokry) [reply source/models/row.py]: ok done

File: source/models/row.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            index = Index(self.table, index_name = index_name, index_value = self.value[index_name])
+            index.update_primary_key(primary_key = self.primary_key) ##update
+    
+    def __get_by_primary_key(self):
+        if not self.row_exists():
+            return None
+        with open(self.get_path(), 'r') as row_file:
+            self.value = json.load(row_file)
+
+    def delete(self):
```
Comment (mohamedabdelsalam) [inline source/models/row.py:50]: It is better to do the following 
```
   if (!row.exist)
        return 
    os.remove...
    .....
```

Comment (hussainmansour) [reply source/models/row.py:50]: i made it

File: source/models/row.py (hunk 3 above)
Comment (josephShokry) [reply source/models/row.py:28]: done

File: source/models/row.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            index.delete_primary_key(primary_key = self.primary_key, index_value = self.value[index_name])
+        self.__unlock()
+
+    def __lock(self):
+        lock_file_path = os.path.join(self.table.get_path(), os.path.join("Lock", self.primary_key + ".json"))
+        try:
+            with open(lock_file_path, "x"):
+                pass
+        except:
+            raise FileNotFound(message = "the primary key is locked now try again later")
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: lock shouldn't throw exception, the concept of lock is waiting till the lock released

Comment (josephShokry) [reply source/models/row.py]: done

File: source/models/row.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            index.delete_primary_key(primary_key = self.primary_key, index_value = self.value[index_name])
+        self.__unlock()
+
+    def __lock(self):
+        lock_file_path = os.path.join(self.table.get_path(), os.path.join("Lock", self.primary_key + ".json"))
+        try:
+            with open(lock_file_path, "x"):
+                pass
+        except:
+            while os.path.isfile(lock_file_path):
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: No, you should try again to lock. 

lock {
 is_lock = false
 while (!is_lock)
   try 
       open file
      is_lock = true
     except
        pass

}

File: source/models/row.py (hunk 3 above)
Comment (mohamedabdelsalam) [reply source/models/row.py:28]: why u didn't solve this ?

File: source/models/row.py (hunk 4 above)
Comment (mohamedabdelsalam) [reply source/models/row.py:50]: why u didn't solve this ?

File: source/models/row.py (hunk 6 above)
Comment (josephShokry) [reply source/models/row.py]: ok, I see that my problem is when the while loop finishes it doesn't make the lock, isn't it??

Comment (josephShokry) [reply source/models/row.py]: your implementation is good but in order to make our implementation work can we just call the ```__lock``` function again after breaking the loop
```
def __lock(self):
        lock_file_path = os.path.join(self.table.get_path(), os.path.join("Lock", self.primary_key + ".json"))
        try:
            with open(lock_file_path, "x"):
                pass
        except:
            while os.path.isfile(lock_file_path):
                time.sleep(1)
            self.__lock()

```

File: source/models/row.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    def __lock(self):
+        while True:
+            try:
+                with open(self.__get_lock_path(), "x"):
+                    break
+            except:
+                pass
+
+    def __unlock(self):
+        lock_file_path = os.path.join(self.table.get_path(), os.path.join("Lock",self.primary_key + ".json"))
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: use `__get_lock_path `

**Review by mohamedabdelsalam — APPROVED** [review]

**Review by hussainmansour — APPROVED** [review]

**Pull request merged**